### PR TITLE
HOSTEDCP-1088: Hypershift e2e framework

### PR DESCRIPTION
**Pull request by muraee**
**What this PR does / why we need it**:

1. get rid of all cleanup sub tests
1. have a better structure of the sequence of common validations/subtests we execute for each test. Each test will have the following sequence :
a. Test/Validate: validates cluster is operational and in good state (executed automatically)
b. Test/Main: this is the main part of the test, created by the user of the framework
c. Test/Teardown: destroys the cluster and namespace
d. Test/PostTeardown: common validations/subtests for all tests that need to run after the cluster has been destroyed (e.g validateMetrics)
1. make sure we stop the test immediately if it fails, currently we continue execution and do more validations

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

**Comment by enxebre** [comment]
Thanks!
I'm not seeing where ValidateMetricsAreExposed is run in the new workflow?
Can you update the PR desc with goals/problems this PR is addressing?
Can you also provide a snapshot of the new resulting output and now it compares/improve the existing one use cases?
Then I think we'd want to make sure this would be suitable for the more elaborated test cases like upgrade, private clusters... I'm not seeing the SwitchFromPrivateToPublic / SwitchFromPublicToPrivate validations in the new workflow? would that go into the "main" test func?
Also another thing that I'd like to address is to have more granular t.run output for the steps within ValidatePrivateCluster, e.g WaitForNodePoolDesiredNodes/WaitForImageRollout/validateHostedClusterConditions so it's immediately clear at a glance at the output which step failed. This can be addressed separately though.

**Comment by muraee** [comment]
> I'm not seeing where ValidateMetricsAreExposed is run in the new workflow?

This would go in tthe `postTeardown()` func, I will add it there, I just didn't want to copy that code yet.

> Then I think we'd want to make sure this would be suitable for the more elaborated test cases like upgrade, private clusters... I'm not seeing the SwitchFromPrivateToPublic / SwitchFromPublicToPrivate validations in the new workflow? would that go into the "main" test func?

yes, any logic specific to a particular test would go in the `Main` test func passed to the `NewHypershifTest()`.
I just changed the `TestCreateCluster` for now to make it clear how the new worklflow works.

**Comment by muraee** [comment]
old output
```xml
<name="TestCreateCluster/EnsureNodeCountMatchesNodePoolReplicas" time="0.040000"/>
<name="TestCreateCluster/EnsureNoCrashingPods" time="0.020000"/>
<name="TestCreateCluster/EnsureNodeCommunication" time="0.120000"/>
<name="TestCreateCluster/EnsurePSANotPrivileged" time="0.130000"/>
<name="TestCreateCluster/EnsureNetworkPolicies/EnsureComponentsHaveNeedManagementKASAccessLabel" time="0.020000"/>
<name="TestCreateCluster/EnsureNetworkPolicies/EnsureEgressTrafficToManagementKAS" time="2.440000"/>
<name="TestCreateCluster/EnsureNetworkPolicies" time="2.470000"/>
<name="TestCreateCluster/EnsureAllRoutesUseHCPRouter" time="0.010000"/>
<name="TestCreateCluster/NoticePreemptionOrFailedScheduling" time="0.020000"/>
<name="TestCreateCluster/EnsureNoPodsWithTooHighPriority" time="0.020000"/>
<name="TestCreateCluster/EnsureHCPContainersHaveResourceRequests" time="0.020000"/>
<name="TestCreateCluster/EnsureAllContainersHavePullPolicyIfNotPresent" time="0.020000"/>
<name="TestCreateCluster/EnsurePodsWithEmptyDirPVsHaveSafeToEvictAnnotations" time="0.020000"/>
<name="TestCreateCluster/PreTeardownClusterDump" time="164.240000"/>
<name="TestCreateCluster/DestroyCluster_1" time="421.060000"/>
<name="TestCreateCluster/EnsureAPIBudget/control-plane-operator_read" time="0.020000"/>
<name="TestCreateCluster/EnsureAPIBudget/control-plane-operator_mutate" time="0.000000"/>
<name="TestCreateCluster/EnsureAPIBudget" time="0.050000"/>
<name="TestCreateCluster/DeleteTestNamespace" time="0.000000"/>
<name="TestCreateCluster/ValidateMetricsAreExposed" time="0.090000"/>
<name="TestCreateCluster" time="1339.690000"/>
```

new output
```xml
<name="TestCreateCluster/Validate/EnsureNodeCountMatchesNodePoolReplicas" time="0.010000"/>
<name="TestCreateCluster/Validate/EnsureNoCrashingPods" time="0.010000"/>
<name="TestCreateCluster/Validate/EnsureNodeCommunication" time="0.060000"/>
<name="TestCreateCluster/Validate/EnsurePSANotPrivileged" time="0.140000"/>
<name="TestCreateCluster/Validate" time="1297.380000"/>
<name="TestCreateCluster/Main" time="0.000000"/>
<name="TestCreateCluster/After/EnsurePodsWithEmptyDirPVsHaveSafeToEvictAnnotations" time="0.020000"/>
<name="TestCreateCluster/After/EnsureAllContainersHavePullPolicyIfNotPresent" time="0.020000"/>
<name="TestCreateCluster/After/EnsureHCPContainersHaveResourceRequests" time="0.010000"/>
<name="TestCreateCluster/After/EnsureNoPodsWithTooHighPriority" time="0.020000"/>
<name="TestCreateCluster/After/NoticePreemptionOrFailedScheduling" time="0.020000"/>
<name="TestCreateCluster/After/EnsureAllRoutesUseHCPRouter" time="0.000000"/>
<name="TestCreateCluster/After/EnsureNetworkPolicies/EnsureComponentsHaveNeedManagementKASAccessLabel" time="0.010000"/>
<name="TestCreateCluster/After/EnsureNetworkPolicies/EnsureEgressTrafficToManagementKAS" time="2.290000"/>
<name="TestCreateCluster/After/EnsureNetworkPolicies" time="2.300000"/>
<name="TestCreateCluster/Teardown" time="551.970000"/>
<name="TestCreateCluster/PostTeardown/EnsureAPIBudget/control-plane-operator_read" time="0.020000"/>
<name="TestCreateCluster/PostTeardown/EnsureAPIBudget/control-plane-operator_mutate" time="0.000000"/>
<name="TestCreateCluster/PostTeardown/ValidateMetricsAreExposed" time="0.050000"/>
<name="TestCreateCluster/PostTeardown" time="0.050000"/>
<name="TestCreateCluster" time="1893.810000"/>
```

File: test/e2e/util/hypershift_framework.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+func (h *hypershiftTest) after(hostedCluster *hyperv1.HostedCluster) {
+	// don't run if test has already failed
+	if h.Failed() {
+		h.Logf("skipping after()")
+		return
+	}
+
+	h.Logf("after()")
+
+	hcpNs := manifests.HostedControlPlaneNamespace(hostedCluster.Namespace, hostedCluster.Name).Name
```
Comment (enxebre) [inline test/e2e/util/hypershift_framework.go:106]: I don't see a reason why this couldn't just go into "before" right after ValidatePublicCluster?
what's this "after" abstraction buying us?

Comment (muraee) [reply test/e2e/util/hypershift_framework.go:106]: mostly to save time on the test.
so we don't wase time running all those validations before the important part of the test `Main` is run, where most of the errors occur. 
so this let us fail as fast as possible.

Comment (enxebre) [reply test/e2e/util/hypershift_framework.go:106]: I see! although it make sense I think the time to run those tests is negligible right? main time penalty on e2e is wait for infra to come up right? I'd be in favour of dropping the abstraction and only add it if the time for those validations happens to be considerable. I don't see a clear semantic diff value between /validate and /after, e.g why does Validate/EnsurePSANotPrivileged belongs to validate? So I'd keep it simple for now, we can reevaluate to be more efficient by grouping by duration or such later on.

File: test/e2e/util/fixture.go [hunk 2]
```diff
@@ -51,7 +42,7 @@ var (
 //
 // This function is intended (for now) to be the preferred default way of
 // creating a hosted cluster during a test.
-func CreateCluster(t *testing.T, ctx context.Context, client crclient.Client, opts *core.CreateOptions, platform hyperv1.PlatformType, artifactDir string, serviceAccountSigningKey []byte) *hyperv1.HostedCluster {
+func _CreateCluster(t *testing.T, ctx context.Context, client crclient.Client, opts *core.CreateOptions, platform hyperv1.PlatformType, artifactDir string, serviceAccountSigningKey []byte) *hyperv1.HostedCluster {
```
Comment (csrwng) [inline test/e2e/util/fixture.go:45]: is this still needed?

**Review by csrwng — COMMENTED** [review]
just a couple of comments
also, will this help us move to 1.20 where we can't do t.Run inside a cleanup function?

File: test/e2e/util/fixture.go (hunk 2 above)
Comment (muraee) [reply test/e2e/util/fixture.go:45]: no, I will be deleting this.

**Comment by enxebre** [comment]
This is great, thanks a lot.

Do we know why this failure timed out https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_hypershift/2834/pull-ci-openshift-hypershift-main-e2e-aws/1686307552859525120?

Follow ups:
- In [this failure](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_hypershift/2834/pull-ci-openshift-hypershift-main-e2e-aws/1685996170989539328) is difficult to grasp why TestCreateClusterCustomConfig and TestUpgradeControlPlane failed. We should have more granular t.runs for those checks so it's clear at a glance.
- Can we change junit to only print the latest level of the nested test https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_hypershift/2834/pull-ci-openshift-hypershift-main-e2e-aws/1686695356840022016?

I think maybe e2eutil.NewHypershiftTest().RunTest() or so it's more accurate than .CreateCluster. And ValidateHostedCluster vs Validate so it's clear when looking at junit for e.g nodepools that those validations are for the HC supporting the test.

/lgtm
/hold

Let's make sure to merge
https://github.com/openshift/hypershift/pull/2775
https://github.com/openshift/hypershift/pull/2792
https://github.com/openshift/hypershift/pull/2848

before this so any unexpected disruption doesn't became a distraction from that work.

**Review by bryan-cox — APPROVED** [review]
/lgtm

**Comment by sjenning** [comment]
waiting for the last of 3 hold PRs to merge https://github.com/openshift/hypershift/pull/2792

**Pull request merged**